### ISSUE TITLE
Docker Compose environment variable precedence is confusing.

### ISSUE THREAD

**Issue by Kelnor277** [docker/docker.github.io#7236]
File: [compose/environment-variables.md](https://docs.docker.com/compose/environment-variables/)

This part:
When you set the same environment variable in multiple files, here’s the priority used by Compose to choose which value to use:

Compose file
Shell environment variables
Environment file
Dockerfile
Variable is not defined

Should be changed to this:

When you set the same environment variable in multiple files, here’s the priority used by Compose to choose which value to use:

Compose file
Environment files specified in the compose file. In reverse order of listing.
Shell environment variables
The .env file selected by docker (I'm not sure what the exact rules are for selecting the .env file.)
Dockerfile
Variable is not defined

By reading what's currently in the docs. I believed that a specified env-file specified in the docker-compose file would be overridable by shell variables.

**Comment by garethr** [comment]
cc. @chris-crone

**Comment by chris-crone** [comment]
> Compose file
> Environment files specified in the compose file. In reverse order of listing.

I don't think that we need to add this to the list as environment files imported in the Compose file are essentially part of the Compose file. The ordering is outlined in the [Compose file documentation](https://docs.docker.com/compose/compose-file/#env_file). If there's continued confusion here, it might be worth linking to that documentation.

> Oh and the special case where the compose file specifies a blank variable: ... Then apparently shell will override, but env file will be ignored?

The [documentation](https://docs.docker.com/compose/environment-variables/#pass-environment-variables-to-containers) is accurate in this case, stating:

```
The value of the DEBUG variable in the container is taken from the value for the same variable in the shell in which Compose is run.
```

It might not be expected behaviour in which case it could be a Compose issue. @shin-  Do you know the reason for the above behaviour?

**Comment by Kelnor277** [comment]
The last example I gave was a case where you if specify the variable and set it to blank with `Var:` not simply specifying the var with no colon.
IE for:
```
environment:
  Var:
```
The value will be overridden (Or at least it was in the version I used when I submitted the issue) if you had a shell variable of the same name. 
So, you've specified an empty string in compose (should be the highest priority,) but it'll get overridden by shell. If you specify a non-empty string (ie `Var:1`) then it's highest priority. Not sure if this was intended, but might be hard to change at this point since people probably rely on this. That's your call.

**Comment by chris-crone** [comment]
@Kelnor277 I didn't catch that subtle difference, thanks for elaborating!

>  but might be hard to change at this point since people probably rely on this

Yes, it would likely be a bad idea to change behaviour like this but good to document it.

**Comment by L-Hudson** [comment]
Thank you all for the discussion; moving this to Docs backlog to be addressed.
